ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/metrics/Metric.scala

```diff
@@ -7,37 +7,51 @@ import com.google.protobuf.timestamp.Timestamp
 
 import java.time.{Clock, Duration, Instant}
 
-sealed trait Metric[T] {
+trait Metric[Elem] {
 
-  def onNext(value: T): Metric[T]
+  type Value <: Metric.MetricValue
 
-  def periodicValue(): (Metric[T], MetricValue)
+  type Objective <: Metric.ServiceLevelObjective[Value]
 
-  def finalValue(totalDurationSeconds: Double): MetricValue
+  def onNext(value: Elem): Metric[Elem]
 
-  def name: String = getClass.getSimpleName
+  def periodicValue(): (Metric[Elem], Value)
 
-}
+  def finalValue(totalDurationSeconds: Double): Value
+
+  def violatedObjectives: Map[Objective, Value] = Map.empty
+
+  def name: String = getClass.getSimpleName
 
-sealed trait MetricValue {
-  def formatted: List[String]
 }
 
 object Metric {
+  trait MetricValue {
+    def formatted: List[String]
+  }
+
+  trait ServiceLevelObjective[MetricValueType <: MetricValue] {
+    def isViolatedBy(metricValue: MetricValueType): Boolean
+    def moreViolatingOf(first: MetricValueType, second: MetricValueType): MetricValueType
```

Comment:
I can't infer the actual purpose of this method. A better name or/and description might help here.

Reply:
I see this is the max function over `MetricValueType`s. Can you extract and use it as such
e.g. 
```
implicit val metricsOrdering: Ordering[MetricValueType] = Ordering.by{...}
```

Reply:
usage example from your test
```
      forAll(cases) { (first, second, expected) =>
        Ordering[MaxDelay].max(first, second) shouldBe expected
      }
```

Reply:
It's sort of max function but in a context of an objective. I want to be able to get the most objective-violating value, so the method is to get a comparison between values. 
My thinking was that we could have an objective for max value, and another on min value, so different comparison functions would be required in different contexts.
Now I think such use case is useless, because in reality we will have only objectives asserting that performance is greater than a threshold and never in the other way.

Reply:
I will implement `Ordering` instances as you suggest 👍

Reply:
I've implemented `Ordering[MaxDelay.Value]` and renamed `def moreViolatingOf()` to `def maxViolatingValue()`.
On seconds thoughts I feel that having this method in `ServiceLevelObjective[T]` trait makes concerns separated more cleanly among classes. A `Metric` knows how to compute metric values based on elements flowing in the stream while a `ServiceLevelObjective` knows which value is not allowed and which value is "better"/"worse" than another.
However, if you think this is unintuitive I can change it to be easier to use.

Reply:
I hate being too opinionated on random stuff, but here I think this method should follow the typeclass pattern. It is just a `Ordering[MaxDelay.Value]` that can be instantiated in e.g. `MaxDelay` companion object and brought into scope wherever you need to perform the `max` or other ordering ops. Having it as a prototype of `ServiceLevelObjective` gives me a false sense of cohesion: it is not using any context like its siblings `isViolatedBy` and `formatted`, whereas you only restrict its usage if you already have a `ServiceLevelObjective` object.